### PR TITLE
Training PSNR Model

### PR DESCRIPTION
**Pull request by captain-pool**
Added First Phase of training. #17 
First phase of training consists of warming up the generator by, training the model on L1 Loss, to reduce PSNR.
The variables of this trained generator will be used as initial weights for Phase 2 of Training.

### CODE REVIEW

File: E2_ESRGAN/lib/phase_1.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from functools import partial
+import tensorflow as tf
+from lib import settings, utils, model, dataset
+
+class Model:
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/phase_1.py]: `class Model(object)` for people who might run this with old python. Here and elsewhere.

File: E2_ESRGAN/lib/model.py [hunk 2]
```diff
@@ -6,7 +6,7 @@ class RRDBNet(tf.keras.Model):
   def __init__(
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/model.py]: Re: Your import style at the top: Try not to use `import *` kind of imports. It pollutes the namespace and can cause surprising behavior.

File: E2_ESRGAN/lib/phase_1.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+from functools import partial
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/phase_1.py]: Is this code structure of having a different file per-training phase followed elsewhere? 
Instead you can split your original model file into one module each for G, D & the overall model. And add each training phase as a helper on the main model class - this way its easier to track the behavior of the model across its entire lifetime in the pipeline.

File: E2_ESRGAN/lib/phase_1.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    self.phase_args = settings["train_psnr"]
+    self.dataset = dataset.load_dataset(
+        dataset_args["name"],
+        dataset.scale_down(
+            method=dataset_args["scale_method"],
+            dimension=dataset_args["dimension"]),
+        batch_size=settings["batch_size"],
+        data_dir=kwargs["data_dir"])
+    self.G = model.RRDBNet(out_channel=3)
+    self.iterations = settings["iterations"]
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/phase_1.py]: Try to have `_iterations` kind of naming for attributes that don't need to be public.

File: E2_ESRGAN/lib/phase_1.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+from functools import partial
+import tensorflow as tf
+from lib import settings, utils, model, dataset
+
+class Model:
+  def __init__(self, **kwargs):
+  """ PSNR model training on L1 Loss to warmup the Generator
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/phase_1.py]: Define the acronym first time you use it :-)

File: E2_ESRGAN/lib/utils.py [hunk 6]
```diff
@@ -1,17 +1,24 @@
 import tensorflow as tf
 from lib.settings import settings
 
-def checkpoint(checkpoint, load=False):
+
+def checkpoint(checkpoint, training_phase, load=False, assert_consumed=True):
   """ Saves or Loads checkpoint.
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/utils.py]: Split this into two methods. Saving & loading are two very separate behaviors and it seems weird in terms of modularity to group them together.

File: E2_ESRGAN/lib/utils.py [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
   """
-  dir_ = settings().get("checkpoint_path", "ckpt")
+  dir_ = settings()["checkpoint_path"][training_phase]
   if not load:
     checkpoint.save(file_prefix=dir_)
   else:
-    checkpoint.restore(tf.train.latest_checkpoint(dir_))
+    status = checkpoint.restore(tf.train.latest_checkpoint(dir_))
+    if assert_consumed:
+      satus.assert_consumed()
```
Comment (srjoglekar246) [inline E2_ESRGAN/lib/utils.py]: status? How did this run? :-P

File: E2_ESRGAN/main.py [hunk 8]
```diff
@@ -1,9 +1,32 @@
+import os
 import argparse
+import logging
+from lib import settings, phase_1
+import tensorflow as tf
+
+
+def main(**kwargs):
```
Comment (srjoglekar246) [inline E2_ESRGAN/main.py]: Where is this being called? Or is that for later?

File: E2_ESRGAN/config.yaml [hunk 9]
```diff
@@ -1,7 +1,33 @@
...
+  name: "open_images_v4"
+  scale_method: "bicubic"
+  hr_dimension: 512
+#Training
+# warmup_num_iter: 1000 # Number of steps to train the PSNR model
+batch_size: 128
+iterations: 2
+train_psnr:
+  adam:
+    initial_lr: 0.0002
```
Comment (srjoglekar246) [inline E2_ESRGAN/config.yaml]: Are these taken from the paper? If yes, add a comment saying so.

File: E2_ESRGAN/lib/utils.py (hunk 7 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/utils.py]: oops. I've not yet tested out the code! so, it might have just got through the cracks. After I fix this cloud issue, I'll run these and check

File: E2_ESRGAN/config.yaml (hunk 9 above)
Comment (captain-pool) [reply E2_ESRGAN/config.yaml]: Yep

Comment (captain-pool) [reply E2_ESRGAN/config.yaml]: Sure

File: E2_ESRGAN/lib/phase_1.py (hunk 3 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/phase_1.py]: Okay!

File: E2_ESRGAN/lib/phase_1.py (hunk 4 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/phase_1.py]: Done :)

File: E2_ESRGAN/main.py (hunk 8 above)
Comment (captain-pool) [reply E2_ESRGAN/main.py]: It will be called from `__name__ == '__main__'` block.

File: E2_ESRGAN/lib/utils.py (hunk 6 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/utils.py]: SG. :P Fixed it :)

File: E2_ESRGAN/lib/phase_1.py (hunk 5 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/phase_1.py]: Done :)

File: E2_ESRGAN/lib/phase_1.py (hunk 3 above)
Comment (captain-pool) [reply E2_ESRGAN/lib/phase_1.py]: I think, this should be kept for later, once all the basic PRs gets merged. Else, it will be difficult to restructure the code, since there some unmerged branches, which are dependent on each other. Once all of them gets merged, we can take a look into this. What do you say?
I've opened an issue to keep track of this. #30

Comment (srjoglekar246) [reply E2_ESRGAN/lib/phase_1.py]: TBH, "it will cause merge conflicts" is never a good reason to keep the code in a bad state. In fact, its a non-trivial chunk of the effort we put into getting in code at Google :-).

The bigger reason why I want a refactor, is because the `Model` class in this file actually trains the `RRDBNet` object from `model.py`. In essence, we now have two classes that encapsulate the same essential object but do two actions on them (which kind-of feels against the spirit of OOP).

As a middle ground, might I suggest:
* Refactor the class in this file to just a `warmup_generator(...)` function that takes in a `RRDBNet` object along with settings etc, and trains it as required.
* Add a `TODO(captain-pool)` block to eventually merge this into the main class.

This is mainly also because all of the the members in the `Model` class in this file are only utilized & accessed in the `train` function, and never elsewhere (or ever again)- so there is not real need for a class.

Comment (captain-pool) [reply E2_ESRGAN/lib/phase_1.py]: SG
